pgxn/neon/neon.c

```diff
@@ -366,23 +366,38 @@ RestoreRunningXactsFromClog(CheckPoint *checkpoint, TransactionId **xids, int *n
 	if (!TransactionIdIsValid(checkpoint->oldestActiveXid))
 	{
 		elog(LOG, "cannot restore running-xacts from CLOG because oldestActiveXid is not set");
-		goto fail;
-	}
+		if (running_xacts_overflow_policy == OP_WAIT)
+			goto fail;
 
-	/*
-	 * We will scan the CLOG starting from the oldest active XID.
-	 *
-	 * In some corner cases, the oldestActiveXid from the last checkpoint
-	 * might already have been truncated from the CLOG. That is,
-	 * oldestActiveXid might be older than oldestXid. That's possible because
-	 * oldestActiveXid is only updated at checkpoints. After the last
-	 * checkpoint, the oldest transaction might have committed, and the CLOG
-	 * might also have been already truncated. So if oldestActiveXid is older
-	 * than oldestXid, start at oldestXid instead. (Otherwise we'd try to
-	 * access CLOG segments that have already been truncated away.)
-	 */
-	from = TransactionIdPrecedes(checkpoint->oldestXid, checkpoint->oldestActiveXid)
-		? checkpoint->oldestActiveXid : checkpoint->oldestXid;
+		if (TransactionIdIsValid(checkpoint->oldestXid))
+		{
+			/* Scan from the beginning of CLOG */
+			from = checkpoint->oldestXid;
+		}
+		else
+		{
+			/* Assume that there are no active transactions except prepared */
+			PrescanPreparedTransactions(xids, nxids);
```

Comment:
Discussed with @knizhnik that he will try to make a repro test based on `test_replica_start_with_too_many_unused_xids` to exercise the new path

Reply:
Right now the problem can not be reproduced with test: oldfstActiveXid is assigned when checkpoint record is replayed (online or shutdown). And checkpoint record is written by initdb. So when we just creates database, we already has checkpoint with valid oldestActiveXid.

So how this problem (with invalid `oldestActiveXid`) can happen at all at prod?
Shutdown checkpoint contains `oldestActiveXid=0` (just because there are no active transactions after shutdown.
Now pageserver contains the following logic:
```
                   // A shutdown checkpoint has `oldestActiveXid == InvalidTransactionid`,
                    // because at shutdown, all in-progress transactions will implicitly
                    // end. Postgres startup code knows that, and allows hot standby to start
                    // immediately from a shutdown checkpoint.
                    //
                    // In Neon, Postgres hot standby startup always behaves as if starting from
                    // an online checkpoint. It needs a valid `oldestActiveXid` value, so
                    // instead of overwriting self.checkpoint.oldestActiveXid with
                    // InvalidTransactionid from the checkpoint WAL record, update it to a
                    // proper value, knowing that there are no in-progress transactions at this
                    // point, except for prepared transactions.
                    //
                    // See also the neon code changes in the InitWalRecovery() function.
                    if xlog_checkpoint.oldestActiveXid == pg_constants::INVALID_TRANSACTION_ID
                        && info == pg_constants::XLOG_CHECKPOINT_SHUTDOWN
                    {
                        let mut oldest_active_xid = self.checkpoint.nextXid.value as u32;
                        for xid in modification.tline.list_twophase_files(lsn, ctx).await? {
                            if (xid.wrapping_sub(oldest_active_xid) as i32) < 0 {
                                oldest_active_xid = xid;
                            }
                        }
                        self.checkpoint.oldestActiveXid = oldest_active_xid;
                    } else {
                        self.checkpoint.oldestActiveXid = xlog_checkpoint.oldestActiveXid;
                    }
```
But this code was added not so long time ago.
May be last time primary was active (and write shutdown checkpoint) before this change was deployed to prod.
In this case last written checkpoint will contain invalid (0) `oldestActiveXid`.

In any case, if it happen with one client, it may also happen with other clients.
This is why I think it will be useful to add logic which can handle such situation.